### PR TITLE
Make the ExpressionCompiler assemblies portable

### PR DESCRIPTION
**Pull request by amcasey**
1. Compile them against Profile7.
   * Swap out a few APIs (e.g. using GetTypeInfo for reflection).
   * Introduce an abstraction layer so that we don't have to instantiate
   DkmDataItem (caused a VerificationException).
   * Include WindowsProxy.winmd as an EmbeddedResource directly, rather
   than through a resx file.
2. Swap the output locations of the ResultProvider projects: Portable goes
to the main output directory and NetFX20 goes to a subdirectory (NetFX20).
3. Include the Portable ResultProvider assemblies in the EE vsix, rather
than the NetFX20 ones.

### CODE REVIEW

**Comment by amcasey** [comment]
FYI @cston @KevinH-MS @ManishJayaswal

**Comment by jaredpar** [comment]
:+1: 

Is there a reason these must be portable vs. targeting CoreCLR instead?

**Comment by tmat** [comment]
Portable7 is targeting CoreCLR. It's a subset. Since we don't need anything that isn't in Profile7 we don't need additional CoreCLR nugets.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpInstructionDecoder.cs [hunk 1]
```diff
@@ -129,19 +129,19 @@ internal override ImmutableArray<TypeParameterSymbol> GetAllTypeParameters(Metho
         internal override CSharpCompilation GetCompilation(DkmClrModuleInstance moduleInstance)
         {
             var appDomain = moduleInstance.AppDomain;
-            var previous = appDomain.GetDataItem<CSharpMetadataContext>();
+            var previous = appDomain.GetDataItem<MetadataContextItem<CSharpMetadataContext>>();
             var metadataBlocks = moduleInstance.RuntimeInstance.GetMetadataBlocks(appDomain);
 
             CSharpCompilation compilation;
-            if (metadataBlocks.HaveNotChanged(previous))
+            if (previous.MetadataContext.Matches(metadataBlocks))
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpInstructionDecoder.cs]: Is 'previous' always non-null?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs [hunk 2]
```diff
@@ -64,7 +64,7 @@ internal string GetErrorMessageAndMissingAssemblyIdentities(DiagnosticBag diagno
         {
             Diagnostic firstError = GetErrorAndMissingAssemblyIdentities(diagnostics, out missingAssemblyIdentities);
             Debug.Assert(firstError != null);
-            return formatter.Format(firstError, preferredUICulture ?? Thread.CurrentThread.CurrentUICulture);
+            return formatter.Format(firstError, preferredUICulture ?? CultureInfo.CurrentUICulture);
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: Should 'preferredUICulture' be set from the Debugger setting?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpInstructionDecoder.cs (hunk 1 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpInstructionDecoder.cs]: Fixed here and in VB.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs (hunk 2 above)
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/EvaluationContextBase.cs]: This change preserves the current behavior.  Feel free to open a bug. :)

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 3]
```diff
@@ -178,9 +175,24 @@ internal static string GetFileNameWithoutExtension(string fileName)
             return fileName;
         }
 
+        private static byte[] GetWindowsProxyBytes()
+        {
+            var assembly = typeof(ExpressionCompiler).GetTypeInfo().Assembly;
+            var resourceName = assembly.GetManifestResourceNames().Single();
+            using (var stream = assembly.GetManifestResourceStream(resourceName))
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Consider hardcoding the resource name.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 4]
```diff
@@ -454,7 +466,8 @@ private static uint GetHResult(Exception e)
 
         internal static string GetUtf8String(this BlobHandle blobHandle, MetadataReader metadataReader)
         {
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Is this method used?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/Placeholders.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace System.Linq
+{
+    internal static class LinqExtensions
+    {
+        public static bool All(this string s, Func<char, bool> predicate)
+        {
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/Placeholders.cs]: Can we use !Enumerable.Any(...) instead?

**Comment by cston** [comment]
LGTM

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 6]
```diff
@@ -178,9 +175,24 @@ internal static string GetFileNameWithoutExtension(string fileName)
             return fileName;
         }
 
+        private static byte[] GetWindowsProxyBytes()
+        {
+            var assembly = typeof(ExpressionCompiler).GetTypeInfo().Assembly;
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Assembly.GetExecutingAssembly()?

Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: GetExecutingAssembly is bad API. Should never be used. I don't think it's portable anyways.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 4 above)
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Apparently not.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 3 above)
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Frankly, I did it this way since I didn't know the name.  I've extracted it from the debugger and hard-coded it.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 6 above)
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Keeping as-is.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs [hunk 7]
```diff
@@ -1,26 +1,35 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Collections.Immutable;
+using System.Linq;
 using Microsoft.CodeAnalysis.ExpressionEvaluator;
 
 namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
 {
-    internal sealed class CSharpMetadataContext : MetadataContext
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs]: Is this change required for portability (i.e. is this just changing to make it a struct instead of a class)?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 8]
```diff
@@ -178,9 +175,23 @@ internal static string GetFileNameWithoutExtension(string fileName)
...
+            using (var stream = assembly.GetManifestResourceStream("Microsoft.CodeAnalysis.ExpressionEvaluator.Resources.WindowsProxy.winmd"))
+            {
+                var bytes = new byte[stream.Length];
+                using (var memoryStream = new MemoryStream(bytes))
+                {
+                    stream.CopyTo(memoryStream);
+                }
+                return bytes;
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Should we try to cache this byte[]?  (re-reading seems expensive)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs [hunk 9]
```diff
@@ -1,26 +1,35 @@
...
 
         internal CSharpMetadataContext(EvaluationContext evaluationContext)
-            : base(evaluationContext.MetadataBlocks)
         {
+            this.MetadataBlocks = evaluationContext.MetadataBlocks;
             this.Compilation = evaluationContext.Compilation;
             this.EvaluationContext = evaluationContext;
         }
+
+        internal bool Matches(ImmutableArray<MetadataBlock> metadataBlocks)
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs]: "Matches" -> simpler *and* better... :smile:

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs (hunk 7 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs]: If it's just refactoring to make struct, consider whether the additional indirection is worth the anticipated cost of the allocations (I assume we don't usually allocate more than a few of these per debug session).  At least for me, the extra indirection makes the code dealing with this type harder to read.

Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs]: The key change here is that there's now a layer between this type and DkmDataItem.  We can't instantiate that type at runtime since we only have a reference assembly.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 8 above)
Comment (amcasey) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: I believe this will only happen once per evaluation context and we already attempt to cache those.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs (hunk 7 above)
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CSharpMetadataContext.cs]: I see--got it.

**Comment by KevinH-MS** [comment]
LGTM

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContextItem.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.Debugger;
+
+namespace Microsoft.CodeAnalysis.ExpressionEvaluator
+{
+    internal sealed class MetadataContextItem<TMetadataContext> : DkmDataItem
+        where TMetadataContext : struct
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataContextItem.cs]: Why this constraint is required?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 6 above)
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Once I heard a scary story about a bug that was caused by cross-assembly inlining of a method containing `Assembly.GetExecutingAssembly()` invocation.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 8 above)
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Does `MemoryStream` guarantee to preserve the content of the underlying array on `Dispose()`?

File: src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicExpressionCompiler.vb [hunk 11]
```diff
@@ -71,15 +71,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
                 ilOffset,
                 localSignatureToken)
 
-			If (previous Is Nothing OrElse context IsNot previous.EvaluationContext) Then
-				appDomain.SetDataItem(DkmDataCreationDisposition.CreateAlways, New VisualBasicMetadataContext(context))
-			End If
+            If (previous Is Nothing OrElse context IsNot previous.MetadataContext.EvaluationContext) Then
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicExpressionCompiler.vb]: Parentheses are not necessary.

File: src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicInstructionDecoder.vb [hunk 12]
```diff
@@ -82,15 +82,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
...
         Friend Overrides Function GetCompilation(moduleInstance As DkmClrModuleInstance) As VisualBasicCompilation
             Dim appDomain = moduleInstance.AppDomain
-            Dim previous = appDomain.GetDataItem(Of VisualBasicMetadataContext)()
+            Dim previous = appDomain.GetDataItem(Of MetadataContextItem(Of VisualBasicMetadataContext))()
             Dim metadataBlocks = moduleInstance.RuntimeInstance.GetMetadataBlocks(appDomain)
 
             Dim compilation As VisualBasicCompilation
-            If metadataBlocks.HaveNotChanged(previous) Then
-                compilation = previous.Compilation
+            If previous IsNot Nothing AndAlso previous.MetadataContext.Matches(metadataBlocks) Then
```
Comment (VladimirReshetnikov) [inline src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicInstructionDecoder.vb]: `previous?.MetadataContext.Matches(metadataBlocks)`
